api/services/bulk-docs.js

```diff
@@ -35,35 +34,69 @@ const generateBatches = (docs, batchSize) => {
   return batches;
 };
 
-const generateBatchPromise = (batch, res, { isFinal } = {}) => {
-  return db.medic.bulkDocs(batch).then(result => {
-    let resString = JSON.stringify(result);
-    resString += isFinal ? '' : ',';
-    res.write(resString);
-  });
+const getDocsToModify = ids => {
+  let docsToDelete;
+  return db.medic.allDocs({ keys: ids, include_docs: true })
+    .then(result => {
+      docsToDelete = markAsDeleted(result);
+      return utils.updateParentContacts(docsToDelete);
+    })
+    .then(docsToUpdate => {
+      const allDocs = docsToDelete.concat(docsToUpdate);
+      checkForDuplicates(allDocs);
+      return allDocs;
+    });
 };
 
-module.exports = {
-  bulkDelete: (docs, res, { batchSize = 100 } = {}) => {
-    let docsToDelete;
-    const keys = docs.map(doc => doc._id);
-    return db.medic.allDocs({ keys, include_docs: true })
-      .then(result => {
-        docsToDelete = markAsDeleted(result);
-        return utils.updateParentContacts(docsToDelete);
+const deleteBatches = (batches, res, docsToRetry, attemptNumber) => {
+  return batches.reduce((promise, batch, index) => {
+    const isFinal = index === batches.length - 1;
+    let docsToModify;
+    return promise
+      .then(() => {
+        return getDocsToModify(batch);
+      })
+      .then(allDocs => {
+        docsToModify = allDocs;
+        return db.medic.bulkDocs(docsToModify);
       })
-      .then(docsToUpdate => {
-        const allDocs = docsToDelete.concat(docsToUpdate);
-        checkForDuplicates(allDocs);
-        const batches = generateBatches(allDocs, batchSize);
+      .then(result => {
+        const errors = result.map(docUpdate => docUpdate.error);
+        const errorIds = docsToModify
+          .filter((doc, index) => errors[index])
+          .map(doc => doc._id);
+        errorIds.forEach(id => docsToRetry.push(id));
 
-        res.type('application/json');
-        res.setHeader('X-Content-Type-Options', 'nosniff');
-        res.write('[');
+        let resString = JSON.stringify(result);
+        if ((!isFinal || docsToRetry.length) && attemptNumber !== 3) {
+          resString += ',';
+        }
+        res.write(resString);
+      });
+  }, Promise.resolve());
+};
 
-        return batches.reduce((promise, batch, index) => {
-          return promise.then(() => generateBatchPromise(batch, res, { isFinal: index === batches.length - 1 }));
-        }, Promise.resolve([]));
+module.exports = {
+  bulkDelete: (docs, res, { batchSize = 100 } = {}) => {
+    const ids = docs.map(doc => doc._id);
+    const batches = generateBatches(ids, batchSize);
+    let docsToRetry = [];
+    res.type('application/json');
+    res.setHeader('X-Content-Type-Options', 'nosniff');
+    res.write('[');
+    return deleteBatches(batches, res, docsToRetry, 0)
+      .then(() => {
+        // Retry any failures a maximum of 3 times
+        return [1, 2, 3].reduce((promise, attemptNumber) => {
```

Comment:
This approach feels a bit unclear to me. Even if nothing is wrong it's still going to fold 4 promises into themselves, which feels misleading.

I think this chunk of code could definitely be iterated on to be clearer :-)